include/wil/cppwinrt_authoring.h

```diff
@@ -330,5 +330,117 @@ struct single_threaded_notifying_property : single_threaded_rw_property<T>
+    using Xaml_PropertyMetadata = winrt::Windows::UI::Xaml::PropertyMetadata;
+#endif
+
+    using Xaml_DependencyObject_GetValue = std::function<winrt::Windows::Foundation::IInspectable(Xaml_DependencyProperty const&)>;
+    using Xaml_DependencyObject_SetValue = std::function<void(Xaml_DependencyProperty const&, winrt::Windows::Foundation::IInspectable const&)>;
+
+    template <typename PropertyType, typename OwnerType>
+    inline Xaml_DependencyProperty register_dependency_property(
+        const std::wstring_view& propertyNameString)
+    {
+        return Xaml_DependencyProperty::Register(
+            propertyNameString,
+            winrt::template xaml_typename<PropertyType>(),
+            winrt::template xaml_typename<OwnerType>(),
+            nullptr);
+    }
+
+    template <typename PropertyType, typename OwnerType, typename DefaultValueType = PropertyType>
+    inline Xaml_DependencyProperty register_dependency_property(
+        const std::wstring_view& propertyNameString,
+        const DefaultValueType& defaultValue,
+        const Xaml_PropertyChangedCallback& propertyChangedCallback = nullptr)
+    {
+        // TODO: assert T and PropertyType are compatible
+        return Xaml_DependencyProperty::Register(
+            propertyNameString,
+            winrt::template xaml_typename<PropertyType>(),
+            winrt::template xaml_typename<OwnerType>(),
+            Xaml_PropertyMetadata{winrt::box_value(defaultValue), propertyChangedCallback});
+    }
+}
+
+//template <typename T>
+//struct single_threaded_dependency_property
+//{
+//    using Type = T;
+//
+//    single_threaded_dependency_property(
+//        wil::details::Xaml_DependencyProperty const& dp,
+//        wil::details::Xaml_DependencyObject_GetValue const& getValue,
+//        wil::details::Xaml_DependencyObject_SetValue const& setValue) :
+//        m_dp(dp), m_getValue(getValue), m_setValue(setValue)
+//    {
+//    }
+//
+//    template <typename Q>
+//    auto& operator()(Q&& q)
+//    {
+//        return winrt::unbox_value<Type>(m_getValue(m_dp));
+//    }
+//
+//    template <typename Q>
+//    auto& operator=(Q&& q)
+//    {
+//        m_setValue(m_dp, winrt::box_value(value));
+//    }
+//
+//private:
+//    wil::details::Xaml_DependencyProperty const& m_dp{ nullptr };
+//    wil::details::Xaml_DependencyObject_GetValue m_getValue{nullptr};
+//    wil::details::Xaml_DependencyObject_SetValue m_setValue{nullptr};
+//};
+
+#define WIL_DEFINE_DP(baseClass, type, name) \
+    static wil::details::Xaml_DependencyProperty name##Property() \
+    { \
+        static wil::details::Xaml_DependencyProperty s_##name##Property = \
+            wil::details::register_dependency_property<type, baseClass>(L"" #name); \
+        return s_##name##Property; \
+    } \
+    auto name() const \
+    { \
+        return winrt::unbox_value<type>(GetValue(name##Property())); \
+    } \
+    void name(type value) const \

```

Comment:
It seems a bit odd that the setter is marked `const`.  I guess it technically is because the pointer to the DP is not changing.

Reply:
I recommend not marking it `const` yeah - it would be extremely odd and break typical assumptions to have a `const foo` but being able to set properties, even if technically it's possible.